### PR TITLE
c10/util/overflows.h: fix float->int range check at wide-integer boundary

### PR DESCRIPTION
**Pull request by r-barnes**
## Summary

The floating-source overload of `c10::overflows` compared against `limit::max()`, which for wide integer types is **not exactly representable** in floating point: `int64` max (`2^63-1`) rounds up to `2^63` when promoted, so `f > limit::max()` was `false` for `f == 2^63` and the value slipped through the check, then became `INT64_MIN` via `static_cast`. This affects `int64`/`uint64` from `double` and `int32`/`int64` from `float`.

Concretely, before this change:
```cpp
double f = 9223372036854775808.0;      // 2^63 == INT64_MAX + 1
c10::overflows<int64_t>(f);            // returned false (bug)
static_cast<int64_t>(f);               // -9223372036854775808  (silent wrap)
```

**Fix:** for integer targets, compare against the exactly-representable upper bound `max() + 1 == 2^digits` (`f >= 2^digits`) instead of `f > max()`. `lowest()` is `0` or a negated power of two and stays exact, so the low side is unchanged. This also fixes `c10::Scalar::equal`, which calls `overflows` directly.

## Test

Adds `c10/test/util/overflows_test.cpp`:
- Wide-int boundaries (`2^63 -> int64`, `2^64 -> uint64`, `2^31 -> int32/float`), narrow ints, NaN/inf, and cast-exactness.
- A `NoSilentWrapSweep` that asserts the core invariant across all integer targets over ~200k probe values: whenever `overflows()` reports a value in range, the subsequent `static_cast` does not silently wrap (round-trip differs by < 1).

Validated standalone (`g++ -std=c++20 -Wall -Wextra`): boundary cases + the ~200k-probe invariant sweep pass with 0 failures. `lintrunner -a` clean.

Authored with the assistance of Claude Code.

### CODE REVIEW

File: c10/test/util/overflows_test.cpp [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+    const double back = static_cast<double>(static_cast<To>(f));
+    EXPECT_LT(std::abs(back - f), 1.0)
+        << "silent wrap converting " << f << " to a "
+        << std::numeric_limits<To>::digits << "-digit integer";
+  }
+}
+
+template <typename To>
+void sweepNoSilentWrap() {
+  const double probes[] = {
```
Comment (Skylion007) [inline c10/test/util/overflows_test.cpp]: Nit: constexpr

File: c10/util/overflows.h [hunk 2]
```diff
@@ -61,13 +61,28 @@ template <typename To, typename From>
...
-  using limit = std::numeric_limits<typename scalar_value_type<To>::type>;
+  using ToScalar = typename scalar_value_type<To>::type;
+  using limit = std::numeric_limits<ToScalar>;
   if (limit::has_infinity && std::isinf(static_cast<double>(f))) {
     return false;
   }
   if (!limit::has_quiet_NaN && (f != f)) {
     return true;
   }
+  if constexpr (std::is_integral_v<ToScalar>) {
```
Comment (Skylion007) [inline c10/util/overflows.h:72]: I do wonder if this behavior should be gated, but we do have std::in_range now...

**Comment by Skylion007** [comment]
@claude is this standard behavior for overflows? Should it be default?

**Review by Skylion007 — APPROVED** [review]
